firedrake/assemble.py

```diff
@@ -20,7 +20,7 @@
 
 
 def assemble(f, tensor=None, bcs=None, form_compiler_parameters=None,
-             inverse=False, nest=None):
+             inverse=False, nest=None, matfree=False, extra_ctx={}):
```

Comment:
There are now two different arguments that control the type of the resulting matrix.  We should merge these into:

```mat_type=["aij", "nest", "matfree"]```

Then there isn't the same thing of "does nest override matfree or not".  The nest argument then disappears from the variational solvers and turns up in the solver_parameters instead.